sklearn/tests/test_calibration.py

```diff
@@ -1077,20 +1081,43 @@ def test_sigmoid_calibration_max_abs_prediction_threshold(global_random_seed):
     assert_allclose(b2, b3, atol=atol)
 
 
-def test_float32_predict_proba(data):
+@pytest.mark.parametrize("use_sample_weight", [True, False])
+@pytest.mark.parametrize("method", ["sigmoid", "isotonic"])
+def test_float32_predict_proba(data, use_sample_weight, method):
     """Check that CalibratedClassifierCV works with float32 predict proba.
 
-    Non-regression test for gh-28245.
+    Non-regression test for gh-28245 and gh-28247.
     """
+    if use_sample_weight:
+        # Use dtype=np.float64 to check that this does not trigger an
+        # unintentional upcasting: the dtype of the base estimator should
+        # control the dtype of the final model.
+        sample_weight = np.ones_like(data[1], dtype=np.float64)
```

Comment:
Just out of curiosity, what happen if it's the other way around? f64 prediction and f32 weight.

Reply:
In short: they must have the same dtype (float32 or float64), it doesn't matter much which one.

Long answer: Those dtypes only matter for the sigmoid calibration. There, the internal loss machinery is used for a stable and fast implementation of the log loss. As it is written in Cython, types matter. In order to not blow up binary size, the dtypes are restricted to float32 and float64, all input arrays (y_observed, y_predicted/raw_predicted, sample_weights) must have the same dtype.

Reply:
I agree with your analysis. Would the above suggestion help make the intent of this clearer?

https://github.com/scikit-learn/scikit-learn/pull/30873/files#r1967416715

Reply:
> Just out of curiosity, what happen if it's the other way around? f64 prediction and f32 weight.

I could also parametrize that, but it seems overkill as the "interesting" case happen when weights and predictions have differing dtypes.